src/planscape/forsys/cluster_stands.py

```diff
@@ -0,0 +1,261 @@
+import numpy as np
+
+from sklearn.cluster import AgglomerativeClustering
+from sklearn.feature_extraction.image import grid_to_graph
+from scipy.sparse.csgraph import connected_components
+
+
+# This compresses stand data by clustering adjacent stands according to a
+# similarity metric.
+# Assumptions:
+#   - stands are arranged as pixels in an image
+#   - although it's ok for the image to be missing pixels, the number of
+#     connected components must be greater than the desired number of clusters
+#
+# Inputs include ...
+#   ... pixel_dist_to_condition_values: a dictionary mapping x-index to y-index
+#           to conditions to scores
+#   ... pixel_width: image width
+#   ... pixel_height: image height
+#   ... priority_condition_max_value: condition scores fall within range,
+#           [0, priority_condition_max_value]
+#   ... priority_weights: weights applied to condition scores before computing
+#           a similarity metric
+#   ... pixel_index_weight:
+#           - weights applied to x-index and y-index values before computing a
+#             similarity metric
+#           - used to control cluster shape - a lower value gives preference to
+#             long strands, a higher value gives preference to smaller circular
+#             clusters
+#   ... num_clusters: the target number of clusters
+#
+# The feature vector used to inform similarity is ...
+# [
+#   pixel_index_weight * x-index,
+#   pixel_index_weight * y-index
+#   normalized_priority_weights ⋅ normalized_priority_conditions
+# ]
+#
+# Normalization is applied to force ...
+#   ... the l2 norm of normalized priority_conditions to 1
+#   ... the l2 norm of normalized priority_weights to a value within the range,
+#       [0, 1]
+# Normalization enables clustering behavior to remain somewhat consistent for a
+#   given pixel_index_weight regardless of priority weight values,
+#   priority condition values, and the number of priority conditions.
+#
+# The output maps cluster ID's to stand index tuples, (x-index, y-index).
+class ClusteredStands():
+    # -------
+    # outputs
+    # -------
+    # map of cluster ID's to stand index tuples.
+    clusters_to_stands: dict[int, tuple[int, int]]
+    # string message - set to None if clustering was successful.
+    cluster_status_message: str | None
+
+    # ----------------------
+    # intermediate variables
+    # ----------------------
+    # normalized priority condition scores.
+    _normalized_pixel_dist_to_condition_values: dict[int,
+                                                     dict[int,
+                                                          dict[str, float]]]
+    # priority weights l2-normalized to a unit vector.
+    _normalized_priority_weights: dict[str, float]
+
+    # NxM matrix denoting whether a pixel is included as a stand.
```

Comment:
can you indicate the keys of each sub collection?

Reply:
Done.